### PR TITLE
Email verification

### PR DESCRIPTION
**Pull request by setar92**
1. Function for sending activating link to user email.
2. Route for verification mail.
3. Route for sending verification link again.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -150,7 +150,7 @@ You should use `.env.example` directory as a reference and consult coaches or co
 
 2. Install dependencies (`node_modules`). Run `yarn install:all` in the root folder.
 
-3. Run the database. You can either run it in docker using command `cd ./docker/autoline && docker-compose -f docker-compose.services.yml up --build` or by installing `Mariadb` on your computer. Docker variant is preferred.
+3. Run the database. You can either run it in docker using command `cd .docker/autoline && docker-compose -f docker-compose.services.yml up --build` or by installing `Mariadb` on your computer. Docker variant is preferred.
```
Comment (RomanMahola) [inline README.md]: Don`t edit README.md

File: backend/prisma/migrations/20220810085748_user_security/migration.sql [hunk 2]
```diff
@@ -0,0 +1,19 @@
+-- CreateTable
```
Comment (RomanMahola) [inline backend/prisma/migrations/20220810085748_user_security/migration.sql]: Drop

File: backend/src/common/enums/app/route-path.enum.ts [hunk 3]
```diff
@@ -0,0 +1,6 @@
+enum RoutPath {
+  MAIL_ACTIVATE = '/activate/:link',
```
Comment (RomanMahola) [inline backend/src/common/enums/app/route-path.enum.ts]: Don`t use verbs in routes

File: backend/src/common/enums/app/route-path.enum.ts [hunk 4]
```diff
@@ -0,0 +1,6 @@
...
+  MAIL_ACTIVATE_AGAIN = '/again/activate',
```
Comment (RomanMahola) [inline backend/src/common/enums/app/route-path.enum.ts]: delete again 
rename activate for 
user/verififcation
user/verification/:link

File: backend/src/controllers/mail_varification/send_mail_again.ts [hunk 5]
```diff
@@ -0,0 +1,35 @@
+import * as userSecurityService from '@services/mail_verification/user_security.service';
+
+import { sendMail } from '../../services/mail_verification/send_mail/sendLink';
```
Comment (RomanMahola) [inline backend/src/controllers/mail_varification/send_mail_again.ts]: use @services no ../..

File: backend/src/controllers/mail_varification/send_mail_again.ts [hunk 6]
```diff
@@ -0,0 +1,35 @@
+import * as userSecurityService from '@services/mail_verification/user_security.service';
+
```
Comment (RomanMahola) [inline backend/src/controllers/mail_varification/send_mail_again.ts]: Delete space

File: backend/src/controllers/mail_varification/send_mail_again.ts [hunk 7]
```diff
@@ -0,0 +1,35 @@
...
+import { sendMail } from '../../services/mail_verification/send_mail/sendLink';
+
```
Comment (RomanMahola) [inline backend/src/controllers/mail_varification/send_mail_again.ts]: Delete space

File: backend/src/controllers/mail_varification/send_mail_again.ts [hunk 8]
```diff
@@ -0,0 +1,35 @@
...
+import type { TypedRequestBody } from '@common/types/controller/controller';
+import type { Response, NextFunction } from 'express';
+
+type SendMailAgainBody = {
```
Comment (RomanMahola) [inline backend/src/controllers/mail_varification/send_mail_again.ts]: EmailBody

File: backend/src/controllers/mail_varification/send_mail_again.ts [hunk 9]
```diff
@@ -0,0 +1,35 @@
...
+  email: string;
+};
+
+const sendMailAgain = async (
```
Comment (RomanMahola) [inline backend/src/controllers/mail_varification/send_mail_again.ts]: delete again again

File: backend/src/controllers/mail_varification/send_mail_again.ts [hunk 10]
```diff
@@ -0,0 +1,35 @@
...
+  req: TypedRequestBody<SendMailAgainBody>,
+  res: Response,
+  next: NextFunction,
+): Promise<void> => {
+  try {
+    const { email } = req.body;
+    const user = await userSecurityService.getUserByEmail(email);
```
Comment (RomanMahola) [inline backend/src/controllers/mail_varification/send_mail_again.ts]: UserService

File: backend/src/controllers/mail_varification/send_mail_again.ts [hunk 11]
```diff
@@ -0,0 +1,35 @@
...
+    if (!user) {
+      return;
+    }
+    const userSecurity = await userSecurityService.getByUserId(user.id);
+    if (!userSecurity?.email_activation_token) {
+      return;
+    }
+    const token = await sendMail(email);
```
Comment (RomanMahola) [inline backend/src/controllers/mail_varification/send_mail_again.ts]: your function violates single responsibility

File: backend/src/controllers/mail_varification/verification.ts [hunk 12]
```diff
@@ -0,0 +1,43 @@
+import * as userSecurityService from '@services/mail_verification/user_security.service';
```
Comment (RomanMahola) [inline backend/src/controllers/mail_varification/verification.ts]: Bad practice

File: backend/src/controllers/mail_varification/send_mail_again.ts [hunk 13]
```diff
@@ -0,0 +1,35 @@
+import * as userSecurityService from '@services/mail_verification/user_security.service';
```
Comment (RomanMahola) [inline backend/src/controllers/mail_varification/send_mail_again.ts]: Bad pracite

File: backend/src/controllers/mail_varification/verification.ts [hunk 14]
```diff
@@ -0,0 +1,43 @@
...
+
+import { MailActivate } from '../../services/mail_verification/send_mail/constants';
+import { validateMailToken } from '../../services/mail_verification/token.service';
```
Comment (RomanMahola) [inline backend/src/controllers/mail_varification/verification.ts]: @services

File: backend/src/controllers/mail_varification/verification.ts [hunk 15]
```diff
@@ -0,0 +1,43 @@
...
+};
+
+const activateMail = async (
+  req: TypedRequestQuery<EmailActivationRequestQuery>,
+  res: Response,
+  next: NextFunction,
+): Promise<void> => {
+  try {
+    const userToken = req.params.link;
+    const user = await userSecurityService.getByToken(userToken);
```
Comment (RomanMahola) [inline backend/src/controllers/mail_varification/verification.ts]: Token not unique

File: backend/src/controllers/mail_varification/verification.ts [hunk 16]
```diff
@@ -0,0 +1,43 @@
...
+    if (!user) {
+      return res.redirect(
+        `${process.env.FRONTEND_URL}${MailActivate.FAILED_URL}`,
```
Comment (RomanMahola) [inline backend/src/controllers/mail_varification/verification.ts]: check import.meta.env

File: backend/src/controllers/mail_varification/verification.ts [hunk 17]
```diff
@@ -0,0 +1,43 @@
...
+      );
+    }
+    const email = validateMailToken(userToken);
+    if (!email) {
+      return res.redirect(
+        `${process.env.FRONTEND_URL}${MailActivate.FAILED_URL}`,
```
Comment (RomanMahola) [inline backend/src/controllers/mail_varification/verification.ts]: check import.meta.env

File: backend/src/controllers/mail_varification/verification.ts [hunk 18]
```diff
@@ -0,0 +1,43 @@
...
+      );
+    }
+    await userSecurityService.removeMailToken(user.id);
+
+    return res.redirect(
+      `${process.env.FRONTEND_URL}${MailActivate.SUCCESS_URL}`,
```
Comment (RomanMahola) [inline backend/src/controllers/mail_varification/verification.ts]: add to const

File: backend/src/routes/auth/auth.router.ts [hunk 19]
```diff
@@ -6,6 +6,7 @@ const PATH = '/auth';
 
 const authRouter = Router();
 
+authRouter.post(`${PATH}/local/signup`, authController.signupLocal);
```
Comment (RomanMahola) [inline backend/src/routes/auth/auth.router.ts]: why local

File: backend/src/services/mail_verification/send_mail/constants.ts [hunk 20]
```diff
@@ -0,0 +1,8 @@
+const MailActivate = {
+  ACTIVATE_URL: 'localhost:3001/api/v1/activate/',
+  SUCCESS_URL: '/mail-activate-success',
+  FAILED_URL: '/mail-activate-failed',
```
Comment (RomanMahola) [inline backend/src/services/mail_verification/send_mail/constants.ts]: `
varification-success
varification-failed
`

File: backend/src/services/mail_verification/send_mail/constants.ts [hunk 21]
```diff
@@ -0,0 +1,8 @@
+const MailActivate = {
+  ACTIVATE_URL: 'localhost:3001/api/v1/activate/',
```
Comment (RomanMahola) [inline backend/src/services/mail_verification/send_mail/constants.ts]: LOCALHOST??????

**USE APP URL**

File: backend/src/services/mail_verification/send_mail/sendLink.ts [hunk 22]
```diff
@@ -0,0 +1,40 @@
+import { ENV } from '@common/enums/app/app';
+import nodemailer from 'nodemailer';
+
+import { generateMailToken } from '../token.service';
+import { MailActivate } from './constants';
+import { getMessage } from './templates/send_verification_link';
+const sendMail = async (email: string): Promise<string> => {
+  const transporter = nodemailer.createTransport({
+    host: ENV.MAILTRAP.EMAIL_HOST,
+    port: 587,
```
Comment (RomanMahola) [inline backend/src/services/mail_verification/send_mail/sendLink.ts]: add to const

File: backend/src/services/mail_verification/send_mail/sendLink.ts [hunk 23]
```diff
@@ -0,0 +1,40 @@
...
+  const options = {
+    from: ENV.MAIL.FROM_EMAIL_VALIDATE,
+    to: email,
+    subject: MailActivate.SUBJECT,
+    html: getMessage(`${MailActivate.ACTIVATE_URL}${token}`),
+  };
+
+  transporter.sendMail(options).catch((err) => {
+    err = {
+      isMailtrap: true,
```
Comment (RomanMahola) [inline backend/src/services/mail_verification/send_mail/sendLink.ts]: Delete

File: backend/src/services/mail_verification/send_mail/sendLink.ts [hunk 24]
```diff
@@ -0,0 +1,40 @@
...
+      message: 'Incorrect email',
```
Comment (RomanMahola) [inline backend/src/services/mail_verification/send_mail/sendLink.ts]: err.message

File: backend/src/services/mail_verification/send_mail/sendLink.ts [hunk 25]
```diff
@@ -0,0 +1,40 @@
...
+    };
+    throw err;
+  });
+
+  return token;
```
Comment (RomanMahola) [inline backend/src/services/mail_verification/send_mail/sendLink.ts]: SIngle responsibility

File: backend/src/services/mail_verification/send_mail/templates/send_verification_link.ts [hunk 26]
```diff
@@ -0,0 +1,31 @@
+export const getMessage = (link: string): string => {
+  return `<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
```
Comment (RomanMahola) [inline backend/src/services/mail_verification/send_mail/templates/send_verification_link.ts]: try to use template

File: backend/src/services/mail_verification/token.service.ts [hunk 27]
```diff
@@ -0,0 +1,34 @@
...
+}
+
+function generateMailToken(payload: EmailTokenPayload): string {
+  const emailToken = jwt.sign(payload, ENV.JWT.SECRET as string, {
+    expiresIn: '7d',
+  });
+  return emailToken;
+}
+
+function validateMailToken(token: string): string | boolean {
```
Comment (RomanMahola) [inline backend/src/services/mail_verification/token.service.ts]: Is it good to return string | boolean

File: backend/src/services/mail_verification/user_security.service.ts [hunk 28]
```diff
@@ -12,6 +12,26 @@ const getByToken = async (
...
+): Promise<Prisma.User_Security | null> => {
+  const userSequrity = await prisma.user_Security.findFirst({
+    where: {
+      user_id: userId,
+    },
+  });
+  return userSequrity;
+};
+
+const getUserByEmail = async (email: string): Promise<Prisma.User | null> => {
```
Comment (RomanMahola) [inline backend/src/services/mail_verification/user_security.service.ts]: provide to User Service

File: frontend/src/common/enums/app/links.enum.ts [hunk 29]
```diff
@@ -0,0 +1,7 @@
+enum AppLinks {
+  ROOT = 'http://localhost:3000/',
+  SIGN_UP = 'http://localhost:3000/sign-up',
+  SIGN_IN = 'http://localhost:3000/sign-in',
```
Comment (RomanMahola) [inline frontend/src/common/enums/app/links.enum.ts]: change localhost:3000

File: frontend/src/components/top-car-card/styles.module.scss [hunk 30]
```diff
@@ -0,0 +1,122 @@
+.container {
```
Comment (RomanMahola) [inline frontend/src/components/top-car-card/styles.module.scss]: Delete this file from this PR

File: frontend/src/components/top-car-card/styles.module.scss [hunk 31]
```diff
@@ -0,0 +1,122 @@
...
+.carLogo {
+  width: 35px;
+  padding-right: 10px;
+}
+
+.carName {
+  color: var(--black);
+  font-weight: 700;
+  font-size: 16px;
+  font-family: Lato, sans-serif;
```
Comment (RomanMahola) [inline frontend/src/components/top-car-card/styles.module.scss]: use variables

File: frontend/src/components/top-car-card/top-car-card.tsx [hunk 32]
```diff
@@ -0,0 +1,58 @@
+import { FC } from 'react';
```
Comment (RomanMahola) [inline frontend/src/components/top-car-card/top-car-card.tsx]: Delete this file from PR

File: yarn.lock [hunk 33]
```diff
@@ -4,19 +4,19 @@
 
 "@babel/code-frame@^7.0.0":
   version "7.18.6"
-  resolved "https://registry.yarnpkg.com/@babel/code-frame/-/code-frame-7.18.6.tgz#3b25d38c89600baa2dcc219edfa88a74eb2c427a"
```
Comment (RomanMahola) [inline yarn.lock]: check

**Review by RomanMahola — CHANGES_REQUESTED** [review]

File: backend/src/controllers/mail_varification/verification.ts (hunk 16 above)
Comment (setar92) [reply backend/src/controllers/mail_varification/verification.ts]: don't work on backend

File: backend/src/controllers/mail_varification/verification.ts (hunk 17 above)
Comment (setar92) [reply backend/src/controllers/mail_varification/verification.ts]: don't work on backend

File: yarn.lock (hunk 33 above)
Comment (setar92) [reply yarn.lock]: My mistake, used "npx prisma migrate" instead "yarn prisma migrate"

File: backend/src/services/mail_verification/send_mail/templates/send_verification_link.ts (hunk 26 above)
Comment (setar92) [reply backend/src/services/mail_verification/send_mail/templates/send_verification_link.ts]: done

File: frontend/src/navigation/routing/routing.tsx [hunk 34]
```diff
@@ -8,7 +8,7 @@ import { Sign } from '../../components/sign/sign';
 
 const Routing: FC = () => {
   const authData = { name: 'Oleksandr' };
-
+
```
Comment (nikita-remeslov) [inline frontend/src/navigation/routing/routing.tsx]: redundant indent

**Review by nikita-remeslov — DISMISSED** [review]
Looks good to me. 1 minor remark needs to be fixed

**Review by nikita-remeslov — APPROVED** [review]

**Pull request merged**